### DISCUSSION TITLE
alos2App ionospheric corrections for pairs with large displacements

### DISCUSSION THREAD

**Discussion by EJFielding** [isce-framework/isce2#676] [Q&A] [answered]
I am using `alos2App.py` to process an ALOS-2 ScanSAR pair that includes a very large earthquake, with several meters of displacement across the fault. It seems that the ionospheric correction module in `alos2App` has some unwrapping error in the upper or lower subband interferograms across the fault that extends across the whole land area of the scene. The ionospheric estimate (see image of the "ion_40rlks_224alks.ion" estimate) shows a large discontinuity across the fault that then causes a bad ionospheric correction. Is there a way to adjust the unwrapping of the subband interferograms to make the ionospheric correction more accurate? Which of the two unwrapped subbands is used in the ionospheric estimation?
 @CunrenLiang 
![ion_40rlks_224alks-ion](https://user-images.githubusercontent.com/7253576/223019965-0d35df7d-37da-4273-8e45-bd8852420ce4.jpg)

When I look at the upper and lower subband interferograms, I don't see an obvious unwrapping error.
After the filtering and interpolation, the final ionospheric correction has a more gradual but still noticeable ramp along the rupture.
![220905-230220_5rlks_28alks-ion-wrap](https://user-images.githubusercontent.com/7253576/223020623-b7969d08-0832-485d-9c02-8a571fa8eb2b.jpg)

**Comment by CunrenLiang** [comment, marked as answer]
in ion/ion_cal
lower_40rlks_224alks.int and upper_40rlks_224alks.int
are the interferograms used in ionosphere computation.

after unwrapping they are
lower_40rlks_224alks.unw and upper_40rlks_224alks.unw

I think there must be phase unwrapping errors in either lower or upper band interferograms, that have caused the problem.


If we use snaphu to unwrap the interferograms, we can set the ruptured areas to zero in the interferograms lower_40rlks_224alks.int and upper_40rlks_224alks.int. Then I believe we can get correctly unwrapped interferograms. This is the trick I usually use to wrap earthquake interferograms in which phase is aliased.

After that, re-run from ion_unwrap.


--------------------------------
Tips on creating a mask file for masking interferogam, coherence etc.

1. create an image (e.g. out.ppm from mdx) from the interferogram that can be viewed by applications such as gimp. Then you can use gimp to mask the difficult areas (see mask_ruptures_gimp.png in the attached package). Create a mask file with values 0 and 1 (such as flag.float in the attached package, you can view it using mdx.py). Note that the mask file must be of the same size (including file width and length) as your interferogram.

to create the mask file,
* open out.ppm in gimp, draw the area to be masked in black as shown in the figure in the package.
* export the file in 'Raw image data' format in gimp.
* use flag.py in the package to create a mask file in ISCE format. Need to change the hard coded file width/length in flag.py

2. Now that we have a mask file, we can use it to set pixels of the interferogram and coherence to zero using imageMath.py in ISCE, e.g.
Use imageMath.py to set pixels to zero.
imageMath.py -e='a*(b!=0)' --a=topophase.flat --b=flag.float -o topophase_mask.flat -t cfloat
#here we have converted coherence or phsig to a two-band file. The two bands are the same.
imageMath.py -e='a*(b!=0);a*(b!=0)' --a=phsig.cor --b=flag.float -o phsig_mask.cor -t float -s BIL


[unwrap.zip](https://github.com/isce-framework/isce2/files/10894786/unwrap.zip)

**Reply by EJFielding** [reply]
Thanks! Do we need to correct the unwrapping of both the upper and lower subband interferograms?

**Reply by CunrenLiang** [reply]
Yes, both should be corrected. They can use the same mask file since they are of same size.

**Reply by CunrenLiang** [reply]
Just noticed we need to comment out this line
https://github.com/isce-framework/isce2/blob/main/components/isceobj/Alos2Proc/runIonUwrap.py#L77

before re-run from ion_unwrap; otherwise it may overwrite manually masked subband interferograms.
